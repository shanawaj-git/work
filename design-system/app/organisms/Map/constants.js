export const MUST_PROVIDE_GOOGLE_KEY = 'Must provide a google-map key';
export const MUST_PROVIDE_WOOSMAP_KEY = 'Must provide a woos-map key';
export const MUST_PROVIDE_MAP_REF = 'Must provide a mapRef';
