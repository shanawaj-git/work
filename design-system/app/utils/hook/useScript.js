import { useEffect, useState } from 'react';

function useScript(src) {
  if (!src) throw new Error('You must provide a src');

  const [loaded, setLoaded] = useState(false);

  useEffect(() => {
    let script = document.querySelector(`script[src="${src}"]`);
    if (!script) {
      script = document.createElement('script');
      script.src = src;
      script.async = true;
      script.onload = () => {
        setLoaded(true);
      };
      document.body.appendChild(script);
    } else {
      setLoaded(true);
    }
  }, [src]);

  return loaded;
}

export default useScript;
