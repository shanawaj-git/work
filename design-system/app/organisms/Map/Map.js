import React, { useState, useRef, useEffect, useMemo } from 'react';
import debounce from 'lodash/debounce';
import PropTypes from 'prop-types';
import useScript from 'utils/hook/useScript';
import { transformedCordinates, uuid } from 'utils/utils';
import {
  MUST_PROVIDE_GOOGLE_KEY,
  MUST_PROVIDE_MAP_REF,
  MUST_PROVIDE_WOOSMAP_KEY,
} from './constants';

const UAE_LAT_LONG = [25.194365, 55.241362];
const ZOOM = 12;
/**
 * List of Events Avaliable on the Map Component
 */
const mapEvents = {
  onBoundsChange: 'bounds_changed',
  onCenterChange: 'center_changed',
  onClick: 'click',
  onDblclick: 'dblclick',
  onDrag: 'drag',
  onDragend: 'dragend',
  onDragstart: 'dragstart',
  onHeadingChange: 'heading_changed',
  onIdle: 'idle',
  onMaptypeidChange: 'maptypeid_changed',
  onMouseMove: 'mousemove',
  onMouseOut: 'mouseout',
  onMouseOver: 'mouseover',
  onRightClick: 'rightclick',
  onZoomChange: 'zoom_changed',
};

const Map = React.forwardRef((props, ref) => {
  // De-structure props
  const {
    children,
    initalCordinates = UAE_LAT_LONG,
    zoom = ZOOM,
    // _ref,
    onCenterChange,
    googleMapKey,
    woosMapKey,
    cordinates,
  } = props;

  // Map Reference from WindowObject
  const [mapWindowRef, setMapWinRef] = useState();
  // Loaded Map Reference
  const [map, setMap] = useState();

  // Map Element Ref
  const mapElement = useRef(null);
  // List of Registered Events
  const activeEvents = useRef({});

  // Switch Map Component between woos-map and google-map
  const MapComponent = googleMapKey ? GoogleMapComponent : WoosMapComponent;

  useEffect(() => {
    if (!mapWindowRef) return undefined;
    const unMount = InitMap();
    return () => {
      unMount();
    };
  }, [mapWindowRef]);

  /**
   * Init the Map.
   * @returns Object : Un-mounter function to unmount  the Map
   */
  const InitMap = () => {
    const LoadedMap =
      mapWindowRef &&
      new mapWindowRef.Map(mapElement.current, {
        center: transformedCordinates(initalCordinates),
        zoom,
        disableDefaultUI: true,
      });

    // eslint-disable-next-line no-param-reassign
    if (ref) ref.current = { map: LoadedMap, mapWindowRef };

    setMap(LoadedMap);

    // Return Unmount Method
    return () => {
      if (mapElement.current) mapElement.current.innerHTML = null;
    };
  };

  useEffect(() => {
    if (!map) return undefined;

    const eventName = Object.keys({ onCenterChange })[0];

    if (!onCenterChange)
      return () => {
        if (activeEvents.current[eventName]) {
          delete activeEvents.current[eventName];
        }
      };

    if (onCenterChange) {
      onCenterChange(undefined, map);
      registerEvent(mapEvents[eventName], onCenterChange);

      return () => {
        if (activeEvents?.current[eventName]) {
          unRegisterEvent(eventName, onCenterChange);
        }
      };
    }

    return null;
  }, [map]);

  /**
   * Register an Event on the Map and save the reference of the event in the events object
   * @param {*} eventName
   * @param {*} callback
   */
  const registerEvent = (eventName, callback) => {
    map.addListener(
      eventName,
      debounce(e => {
        callback(e, map);
      }),

      500,
    );

    activeEvents.current[eventName] = callback;
  };

  /**
   * Un-Register an Event on the Map and remove the reference of the event in the events object
   * @param {*} eventName
   * @param {*} callback
   */
  const unRegisterEvent = (eventName, callback) => {
    if (activeEvents.current[eventName]) {
      delete activeEvents.current[eventName];
    }

    map.removeEventListener(mapEvents[eventName], callback, false);
  };

  useEffect(() => {
    if (!cordinates) return undefined;
    JumpToCordinates();
    return undefined;
  }, [cordinates]);

  // Map Jump to Cordinates
  const JumpToCordinates = () => map?.panTo(transformedCordinates(cordinates));

  const InjectPropsToChildren = _children =>
    React.cloneElement(_children, {
      mapInstance: {
        map,
        mapRef: mapWindowRef,
      },
    });

  const renderChildren = () => {
    // append Map to all the children of the component
    const newChildren = (children?.props?.children || [children]).map(child => {
      const childToMount = child ? InjectPropsToChildren(child) : child;

      return (
        <React.Fragment key={child?.key || uuid()}>
          {childToMount}
        </React.Fragment>
      );
    });

    return newChildren;
  };

  const Markers = useMemo(() => renderChildren(), [map, children]);

  return (
    <MapComponent
      woosMapKey={woosMapKey}
      googleMapKey={googleMapKey}
      mapWindowRef={mapRef => setMapWinRef(mapRef)}
    >
      <div
        ref={mapElement}
        style={{
          height: '100%',
          width: '100%',
        }}
      >
        {Markers}
      </div>
    </MapComponent>
  );
});

export default Map;

export const MapPropsTypes = {
  cordinates: PropTypes.array,
  initalCordinates: PropTypes.array,
  zoom: PropTypes.number,
  onCenterChange: PropTypes.func,
  googleMapKey: PropTypes.string,
  woosMapKey: PropTypes.string.isRequired,
  children: PropTypes.any,
};

Map.propTypes = {
  cordinates: PropTypes.array,
  initalCordinates: PropTypes.array,
  zoom: PropTypes.number,
  onCenterChange: PropTypes.func,
  googleMapKey: PropTypes.string,
  woosMapKey: PropTypes.string.isRequired,
  children: PropTypes.any,
};

export const GoogleMapComponent = ({
  children,
  googleMapKey,
  mapWindowRef,
}) => {
  if (!googleMapKey) throw new Error(MUST_PROVIDE_GOOGLE_KEY);
  if (!mapWindowRef) throw new Error(MUST_PROVIDE_MAP_REF);

  const [map, setMap] = useState(null);
  const googleMapCDN = useScript(
    `https://maps.googleapis.com/maps/api/js?key=${googleMapKey}&libraries=places`,
  );

  useEffect(() => {
    if (!googleMapCDN) return undefined;
    setMap(window.google.maps);
    return () => null;
  }, [googleMapCDN]);

  useEffect(() => {
    if (map && mapWindowRef) {
      mapWindowRef(map);
    }
  }, [map]);
  return <>{children}</>;
};

GoogleMapComponent.propTypes = {
  googleMapKey: PropTypes.string.isRequired,
  children: PropTypes.any,
  mapWindowRef: PropTypes.func,
};

export const WoosMapComponent = ({ children, woosMapKey, mapWindowRef }) => {
  if (!woosMapKey) throw new Error(MUST_PROVIDE_WOOSMAP_KEY);
  if (!mapWindowRef) throw new Error(MUST_PROVIDE_MAP_REF);

  const mapLoaded = useScript(
    `https://sdk.woosmap.com/map/map.js?key=${woosMapKey}&callback=initMap`,
  );

  useEffect(() => {
    if (!mapLoaded) return undefined;
    mapWindowRef(window.woosmap.map);
    return undefined;
  }, [mapLoaded]);

  if (!mapLoaded) return <>Loading ...</>;
  return children;
};
WoosMapComponent.propTypes = {
  woosMapKey: PropTypes.string.isRequired,
  children: PropTypes.any,
  mapWindowRef: PropTypes.func,
};
