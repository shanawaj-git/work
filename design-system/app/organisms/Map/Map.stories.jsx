import React from 'react';
import Map from './Map';
import StoryWrapper from 'molecules/utility/StoryWrapper';

export default {
  title: 'Components/Map/MapComponent',
  component: Map,
  argTypes: {
    themeType: {
      control: { type: 'select', options: ['DEFAULT', 'MPC'] },
    },
    onChange: { table: { disable: true } },
    groupName: { table: { disable: true } },
    defaultSelect: { table: { disable: true } },
  },
};

const Template = args => (
  <StoryWrapper themeVariant={args.themeType}>
    <div style={{ height: 500, widht: 500 }}>
      <Map {...args} />
    </div>
  </StoryWrapper>
);

export const DefaultMap = Template.bind({});

DefaultMap.args = {
  woosMapKey: process.env.WOOS_MAP_KEY,
  googleMapKey: process.env.GOOGLE_MAP_KEY,
  initalCordinates: [25.204849, 55.270782],
  onCenterChange: undefined,
};

export const WoosMap = Template.bind({});

WoosMap.args = {
  woosMapKey: process.env.WOOS_MAP_KEY,
  initalCordinates: [25.204849, 55.270782],
  onCenterChange: undefined,
};
export const GoogeMap = Template.bind({});

GoogeMap.args = {
  woosMapKey: process.env.WOOS_MAP_KEY,
  googleMapKey: process.env.GOOGLE_MAP_KEY,
  initalCordinates: [25.204849, 55.270782],
  onCenterChange: undefined,
};
