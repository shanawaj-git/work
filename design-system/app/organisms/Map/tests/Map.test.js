import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { render, unmountComponentAtNode } from 'react-dom';
import { render as rtlRender } from 'react-testing-library';
import * as useScript from '../../../utils/hook/useScript';
import {
  MUST_PROVIDE_GOOGLE_KEY,
  MUST_PROVIDE_MAP_REF,
  MUST_PROVIDE_WOOSMAP_KEY,
} from '../constants';
import MapComponentIndex from '../index';
import { GoogleMapComponent } from '../Map';

class MapMock {
  load = jest.fn().mockImplementation(fn => fn({ map: {} }));
}

describe('Map.js', () => {
  describe('index', () => {
    window.WoosmapLoader = {
      load: jest
        .fn()
        .mockImplementation(options =>
          options.callback ? options.callback() : options(),
        ),
    };

    let container = null;

    jest.spyOn(React, 'useEffect').mockImplementation(React.useLayoutEffect);

    beforeAll(() => {
      jest.mock('../../../utils/hook/useScript', () => jest.fn());
    });
    afterAll(() => React.useEffect.mockRestore());

    beforeEach(() => {
      // setup a DOM element as a render target
      container = document.createElement('div');
      document.body.appendChild(container);
    });

    afterEach(() => {
      // cleanup on exiting
      unmountComponentAtNode(container);
      container.remove();
      container = null;
    });

    it('Should import succesfully', () => {
      const tree = renderer.create(
        <MapComponentIndex
          woosMapKey={process.env.WOOS_MAP_KEY}
          googleMapKey={process.env.GOOGLE_MAP_KEY}
          initalCordinates={[25.204849, 55.270782]}
        />,
      );

      expect(tree).toMatchSnapshot();
    });
    it('Should throw if no woos-map key', () => {
      expect(() =>
        renderer.create(
          <MapComponentIndex initalCordinates={[25.204849, 55.270782]} />,
        ),
      ).toThrow(MUST_PROVIDE_WOOSMAP_KEY);
    });

    it('Should throw if google key is not provided to GoogleMapComponent', () => {
      expect(() =>
        renderer.create(
          <GoogleMapComponent woosMapKey={process.env.WOOS_MAP_KEY} />,
        ),
      ).toThrow(MUST_PROVIDE_GOOGLE_KEY);
    });

    it('Should throw if map ref is not provided to GoogleMapComponent', () => {
      expect(() =>
        renderer.create(
          <GoogleMapComponent
            googleMapKey={process.env.GOOGLE_MAP_KEY}
            woosMapKey={process.env.WOOS_MAP_KEY}
          />,
          container,
        ),
      ).toThrow(MUST_PROVIDE_MAP_REF);
    });

    it('WoosMap Should be injected in to the DOM', async () => {
      window.woosmap = {
        MapsLoader: MapMock,
        map: {
          Map: MapMock,
        },
      };
      act(() => {
        render(
          <MapComponentIndex
            woosMapKey={process.env.WOOS_MAP_KEY}
            initalCordinates={[25.204849, 55.270782]}
          />,
          container,
        );
      });

      expect(document.querySelector('body')).toMatchSnapshot();

      act(() => {
        render(
          <MapComponentIndex
            woosMapKey={process.env.WOOS_MAP_KEY}
            initalCordinates={[25.204849, 55.270782]}
          />,
          container,
        );
      });

      expect(document.querySelector('body')).toMatchSnapshot();
    });

    // You cant check if map is rendering the children because map is injecting dynamically and always expecting a marker type component, if it didnt find anything related to marker
    it('Render google map with children', async () => {
      useScript.default = jest.fn().mockImplementation(() => true);

      class Map {}
      window.google = {
        maps: Map,
      };

      window.woosmap = {
        MapsLoader: MapMock,
        map: {
          Map: MapMock,
        },
      };

      const elem = () => <span>Content</span>;

      const tree = rtlRender(
        <MapComponentIndex
          woosMapKey={process.env.WOOS_MAP_KEY}
          googleMapKey={process.env.GOOGLE_MAP_KEY}
          cordinates={[25.204849, 55.270782]}
        >
          <elem />
        </MapComponentIndex>,
      );
      tree.rerender(
        <MapComponentIndex
          woosMapKey={process.env.WOOS_MAP_KEY}
          googleMapKey={process.env.GOOGLE_MAP_KEY}
          cordinates={[25.204849, 55.270782]}
        >
          <elem />
        </MapComponentIndex>,
      );

      expect(tree.baseElement).toMatchSnapshot();
      expect(tree.container).toMatchSnapshot();
    });

    it('load woosmap component', async () => {
      useScript.default = jest.fn().mockImplementation(() => true);

      window.woosmap = {
        MapsLoader: MapMock,
        map: {
          Map: MapMock,
        },
      };

      render(
        <MapComponentIndex
          woosMapKey={process.env.WOOS_MAP_KEY}
          initalCordinates={[25.204849, 55.270782]}
        />,
        container,
      );

      render(
        <MapComponentIndex
          woosMapKey={process.env.WOOS_MAP_KEY}
          initalCordinates={[25.204849, 55.270782]}
        />,
        container,
      );

      render(
        <MapComponentIndex
          woosMapKey={process.env.WOOS_MAP_KEY}
          initalCordinates={[25.204849, 55.270782]}
        />,
        container,
      );

      render(
        <MapComponentIndex
          woosMapKey={process.env.WOOS_MAP_KEY}
          initalCordinates={[25.204849, 55.270782]}
        />,
        container,
      );

      expect(useScript.default).toHaveBeenCalled();
    });

    it('unregister map component on unmounting', async () => {
      useScript.default = jest.fn().mockImplementation(() => true);

      window.woosmap = {
        MapsLoader: MapMock,
        map: {
          Map: MapMock,
        },
      };

      const tree = render(
        <MapComponentIndex
          googleMapKey="testtttt2"
          woosMapKey={process.env.WOOS_MAP_KEY}
          initalCordinates={[25.204849, 55.270782]}
        />,
        container,
      );

      render(
        <MapComponentIndex
          googleMapKey="testtttt2"
          woosMapKey={process.env.WOOS_MAP_KEY}
          initalCordinates={[25.204849, 55.270782]}
        />,
        container,
      );

      try {
        tree.unmount();
      } catch (e) {
        console.error(e);
      }

      expect(useScript.default).toHaveBeenCalled();
    });
  });
});
