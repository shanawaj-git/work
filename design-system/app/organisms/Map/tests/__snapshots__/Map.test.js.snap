// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map.js index Render google map with children 1`] = `
<body>
  <script
    src="https://maps.googleapis.com/maps/api/js?key=AIzaSyBn3kw1bNdgmiXAczwr2DcKLAaW-M3nX14&libraries=places"
  />
  <script
    src="https://sdk.woosmap.com/map/map.js?key=woos-48c80350-88aa-333e-835a-07f4b658a9a4&callback=initMap"
  />
  <div />
  <div>
    <div
      style="height: 100%; width: 100%;"
    >
      <elem
        mapinstance="[object Object]"
      />
    </div>
  </div>
</body>
`;

exports[`Map.js index Render google map with children 2`] = `
<div>
  <div
    style="height: 100%; width: 100%;"
  >
    <elem
      mapinstance="[object Object]"
    />
  </div>
</div>
`;

exports[`Map.js index Should import succesfully 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
/>
`;

exports[`Map.js index WoosMap Should be injected in to the DOM 1`] = `
<body>
  <script
    src="https://maps.googleapis.com/maps/api/js?key=AIzaSyBn3kw1bNdgmiXAczwr2DcKLAaW-M3nX14&libraries=places"
  />
  <div>
    Loading ...
  </div>
  <script
    src="https://sdk.woosmap.com/map/map.js?key=woos-48c80350-88aa-333e-835a-07f4b658a9a4&callback=initMap"
  />
</body>
`;

exports[`Map.js index WoosMap Should be injected in to the DOM 2`] = `
<body>
  <script
    src="https://maps.googleapis.com/maps/api/js?key=AIzaSyBn3kw1bNdgmiXAczwr2DcKLAaW-M3nX14&libraries=places"
  />
  <div>
    Loading ...
  </div>
  <script
    src="https://sdk.woosmap.com/map/map.js?key=woos-48c80350-88aa-333e-835a-07f4b658a9a4&callback=initMap"
  />
</body>
`;
